// langchain/dist/llms/index.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Replicate = exports.HuggingFaceInference = exports.Cohere = exports.OpenAIChat = exports.PromptLayerOpenAI = exports.OpenAI = exports.LLM = exports.BaseLLM = void 0;
/* #__PURE__ */ console.error("[WARN] Importing from 'langchain/llms' is deprecated. Import from eg. 'langchain/llms/openai' instead. See https://js.langchain.com/docs/getting-started/install#updating-from-0052 for upgrade instructions.");
var base_js_1 = require("./base.cjs");
Object.defineProperty(exports, "BaseLLM", { enumerable: true, get: function () { return base_js_1.BaseLLM; } });
Object.defineProperty(exports, "LLM", { enumerable: true, get: function () { return base_js_1.LLM; } });
var openai_js_1 = require("./openai.cjs");
Object.defineProperty(exports, "OpenAI", { enumerable: true, get: function () { return openai_js_1.OpenAI; } });
Object.defineProperty(exports, "PromptLayerOpenAI", { enumerable: true, get: function () { return openai_js_1.PromptLayerOpenAI; } });
var openai_chat_js_1 = require("./openai-chat.cjs");
Object.defineProperty(exports, "OpenAIChat", { enumerable: true, get: function () { return openai_chat_js_1.OpenAIChat; } });
var cohere_js_1 = require("./cohere.cjs");
Object.defineProperty(exports, "Cohere", { enumerable: true, get: function () { return cohere_js_1.Cohere; } });
var hf_js_1 = require("./hf.cjs");
Object.defineProperty(exports, "HuggingFaceInference", { enumerable: true, get: function () { return hf_js_1.HuggingFaceInference; } });
var replicate_js_1 = require("./replicate.cjs");
Object.defineProperty(exports, "Replicate", { enumerable: true, get: function () { return replicate_js_1.Replicate; } });
