// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/models/HnswConfig.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Pinecone API
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: version not set
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.HnswConfigToJSON = exports.HnswConfigFromJSONTyped = exports.HnswConfigFromJSON = exports.instanceOfHnswConfig = void 0;
var runtime_1 = require("../runtime");
/**
 * Check if a given object implements the HnswConfig interface.
 */
function instanceOfHnswConfig(value) {
    var isInstance = true;
    return isInstance;
}
exports.instanceOfHnswConfig = instanceOfHnswConfig;
function HnswConfigFromJSON(json) {
    return HnswConfigFromJSONTyped(json, false);
}
exports.HnswConfigFromJSON = HnswConfigFromJSON;
function HnswConfigFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'efConstruction': !(0, runtime_1.exists)(json, 'ef_construction') ? undefined : json['ef_construction'],
        'ef': !(0, runtime_1.exists)(json, 'ef') ? undefined : json['ef'],
        'm': !(0, runtime_1.exists)(json, 'M') ? undefined : json['M'],
        'maxElements': !(0, runtime_1.exists)(json, 'max_elements') ? undefined : json['max_elements'],
    };
}
exports.HnswConfigFromJSONTyped = HnswConfigFromJSONTyped;
function HnswConfigToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'ef_construction': value.efConstruction,
        'ef': value.ef,
        'M': value.m,
        'max_elements': value.maxElements,
    };
}
exports.HnswConfigToJSON = HnswConfigToJSON;
