// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/models/ProtobufNullValue.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Pinecone API
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: version not set
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProtobufNullValueToJSON = exports.ProtobufNullValueFromJSONTyped = exports.ProtobufNullValueFromJSON = exports.ProtobufNullValue = void 0;
/**
 * `NullValue` is a singleton enumeration to represent the null value for the
 * `Value` type union.
 *
 *  The JSON representation for `NullValue` is JSON `null`.
 *
 *  - NULL_VALUE: Null value.
 * @export
 */
exports.ProtobufNullValue = {
    NullValue: 'NULL_VALUE'
};
function ProtobufNullValueFromJSON(json) {
    return ProtobufNullValueFromJSONTyped(json, false);
}
exports.ProtobufNullValueFromJSON = ProtobufNullValueFromJSON;
function ProtobufNullValueFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
exports.ProtobufNullValueFromJSONTyped = ProtobufNullValueFromJSONTyped;
function ProtobufNullValueToJSON(value) {
    return value;
}
exports.ProtobufNullValueToJSON = ProtobufNullValueToJSON;
