// langchain/dist/callbacks/handlers/tracer_langchain.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LangChainTracer = void 0;
const env_js_1 = require("../../util/env.cjs");
const tracer_js_1 = require("./tracer.cjs");
class LangChainTracer extends tracer_js_1.BaseTracer {
    constructor({ exampleId, tenantId, sessionName, sessionExtra, } = {}) {
        super();
        Object.defineProperty(this, "name", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: "langchain_tracer"
        });
        Object.defineProperty(this, "endpoint", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: (typeof process !== "undefined"
                ? // eslint-disable-next-line no-process-env
                    process.env?.LANGCHAIN_ENDPOINT
                : undefined) || "http://localhost:8000"
        });
        Object.defineProperty(this, "headers", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: {
                "Content-Type": "application/json",
            }
        });
        Object.defineProperty(this, "sessionName", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "sessionExtra", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "session", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "exampleId", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "tenantId", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        // eslint-disable-next-line no-process-env
        if (typeof process !== "undefined" && process.env?.LANGCHAIN_API_KEY) {
            // eslint-disable-next-line no-process-env
            this.headers["x-api-key"] = process.env?.LANGCHAIN_API_KEY;
        }
        this.tenantId =
            tenantId ??
                (typeof process !== "undefined"
                    ? // eslint-disable-next-line no-process-env
                        process.env?.LANGCHAIN_TENANT_ID
                    : undefined);
        this.sessionName =
            sessionName ??
                (typeof process !== "undefined"
                    ? // eslint-disable-next-line no-process-env
                        process.env?.LANGCHAIN_SESSION
                    : undefined) ??
                "default";
        this.sessionExtra = sessionExtra;
        this.exampleId = exampleId;
    }
    async ensureSession() {
        if (this.session) {
            return this.session;
        }
        const tenantId = await this.ensureTenantId();
        const endpoint = `${this.endpoint}/sessions?upsert=true`;
        const res = await fetch(endpoint, {
            method: "POST",
            headers: this.headers,
            body: JSON.stringify({
                name: this.sessionName,
                tenant_id: tenantId,
                extra: this.sessionExtra,
            }),
        });
        if (!res.ok) {
            const body = await res.text();
            throw new Error(`Failed to create session: ${res.status} ${res.statusText} ${body}`);
        }
        const session = await res.json();
        this.session = session;
        return session;
    }
    async ensureTenantId() {
        if (this.tenantId) {
            return this.tenantId;
        }
        const endpoint = `${this.endpoint}/tenants`;
        const response = await fetch(endpoint, {
            method: "GET",
            headers: this.headers,
        });
        if (!response.ok) {
            const body = await response.text();
            throw new Error(`Failed to fetch tenant ID: ${response.status} ${response.statusText} ${body}`);
        }
        const tenants = await response.json();
        if (!tenants || tenants.length === 0) {
            throw new Error(`No tenants found for endpoint ${endpoint}`);
        }
        const tenantId = tenants[0].id;
        this.tenantId = tenantId;
        return tenantId;
    }
    async _convertToCreate(run, example_id = undefined) {
        const session = await this.ensureSession();
        const runExtra = run.extra ?? {};
        runExtra.runtime = await (0, env_js_1.getRuntimeEnvironment)();
        const persistedRun = {
            id: run.id,
            name: run.name,
            start_time: run.start_time,
            end_time: run.end_time,
            run_type: run.run_type,
            reference_example_id: example_id,
            extra: runExtra,
            execution_order: run.execution_order,
            serialized: run.serialized,
            error: run.error,
            inputs: run.inputs,
            outputs: run.outputs ?? {},
            session_id: session.id,
            child_runs: await Promise.all(run.child_runs.map((child_run) => this._convertToCreate(child_run))),
        };
        return persistedRun;
    }
    async persistRun(run) {
        const persistedRun = await this._convertToCreate(run, this.exampleId);
        const endpoint = `${this.endpoint}/runs`;
        const response = await fetch(endpoint, {
            method: "POST",
            headers: this.headers,
            body: JSON.stringify(persistedRun),
        });
        if (!response.ok) {
            const body = await response.text();
            throw new Error(`Failed to persist run: ${response.status} ${response.statusText} ${body}`);
        }
    }
}
exports.LangChainTracer = LangChainTracer;
