// ml-distance/lib/similarities.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.kumarHassebrook = exports.tanimoto = exports.squaredChord = exports.pearson = exports.motyka = exports.kulczynski = exports.intersection = exports.dice = exports.czekanowski = exports.cosine = exports.tree = void 0;
const tree = __importStar(require("ml-tree-similarity"));
exports.tree = tree;
var cosine_1 = require("./similarities/cosine");
Object.defineProperty(exports, "cosine", { enumerable: true, get: function () { return __importDefault(cosine_1).default; } });
var czekanowski_1 = require("./similarities/czekanowski");
Object.defineProperty(exports, "czekanowski", { enumerable: true, get: function () { return __importDefault(czekanowski_1).default; } });
var dice_1 = require("./similarities/dice");
Object.defineProperty(exports, "dice", { enumerable: true, get: function () { return __importDefault(dice_1).default; } });
var intersection_1 = require("./similarities/intersection");
Object.defineProperty(exports, "intersection", { enumerable: true, get: function () { return __importDefault(intersection_1).default; } });
var kulczynski_1 = require("./similarities/kulczynski");
Object.defineProperty(exports, "kulczynski", { enumerable: true, get: function () { return __importDefault(kulczynski_1).default; } });
var motyka_1 = require("./similarities/motyka");
Object.defineProperty(exports, "motyka", { enumerable: true, get: function () { return __importDefault(motyka_1).default; } });
var pearson_1 = require("./similarities/pearson");
Object.defineProperty(exports, "pearson", { enumerable: true, get: function () { return __importDefault(pearson_1).default; } });
var squaredChord_1 = require("./similarities/squaredChord");
Object.defineProperty(exports, "squaredChord", { enumerable: true, get: function () { return __importDefault(squaredChord_1).default; } });
var tanimoto_1 = require("./similarities/tanimoto");
Object.defineProperty(exports, "tanimoto", { enumerable: true, get: function () { return __importDefault(tanimoto_1).default; } });
var kumarHassebrook_1 = require("./similarities/kumarHassebrook");
Object.defineProperty(exports, "kumarHassebrook", { enumerable: true, get: function () { return __importDefault(kumarHassebrook_1).default; } });
