// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/models/ApproximatedConfig.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Pinecone API
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: version not set
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ApproximatedConfigToJSON = exports.ApproximatedConfigFromJSONTyped = exports.ApproximatedConfigFromJSON = exports.instanceOfApproximatedConfig = void 0;
var runtime_1 = require("../runtime");
/**
 * Check if a given object implements the ApproximatedConfig interface.
 */
function instanceOfApproximatedConfig(value) {
    var isInstance = true;
    return isInstance;
}
exports.instanceOfApproximatedConfig = instanceOfApproximatedConfig;
function ApproximatedConfigFromJSON(json) {
    return ApproximatedConfigFromJSONTyped(json, false);
}
exports.ApproximatedConfigFromJSON = ApproximatedConfigFromJSON;
function ApproximatedConfigFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'kBits': !(0, runtime_1.exists)(json, 'k_bits') ? undefined : json['k_bits'],
        'hybrid': !(0, runtime_1.exists)(json, 'hybrid') ? undefined : json['hybrid'],
    };
}
exports.ApproximatedConfigFromJSONTyped = ApproximatedConfigFromJSONTyped;
function ApproximatedConfigToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'k_bits': value.kBits,
        'hybrid': value.hybrid,
    };
}
exports.ApproximatedConfigToJSON = ApproximatedConfigToJSON;
