// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/models/IndexMeta.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Pinecone API
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: version not set
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.IndexMetaToJSON = exports.IndexMetaFromJSONTyped = exports.IndexMetaFromJSON = exports.instanceOfIndexMeta = void 0;
var runtime_1 = require("../runtime");
var IndexMetaDatabase_1 = require("./IndexMetaDatabase");
var IndexMetaStatus_1 = require("./IndexMetaStatus");
/**
 * Check if a given object implements the IndexMeta interface.
 */
function instanceOfIndexMeta(value) {
    var isInstance = true;
    return isInstance;
}
exports.instanceOfIndexMeta = instanceOfIndexMeta;
function IndexMetaFromJSON(json) {
    return IndexMetaFromJSONTyped(json, false);
}
exports.IndexMetaFromJSON = IndexMetaFromJSON;
function IndexMetaFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'database': !(0, runtime_1.exists)(json, 'database') ? undefined : (0, IndexMetaDatabase_1.IndexMetaDatabaseFromJSON)(json['database']),
        'status': !(0, runtime_1.exists)(json, 'status') ? undefined : (0, IndexMetaStatus_1.IndexMetaStatusFromJSON)(json['status']),
    };
}
exports.IndexMetaFromJSONTyped = IndexMetaFromJSONTyped;
function IndexMetaToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'database': (0, IndexMetaDatabase_1.IndexMetaDatabaseToJSON)(value.database),
        'status': (0, IndexMetaStatus_1.IndexMetaStatusToJSON)(value.status),
    };
}
exports.IndexMetaToJSON = IndexMetaToJSON;
