// langchain/dist/vectorstores/index.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PrismaVectorStore = exports.SupabaseVectorStore = exports.SaveableVectorStore = exports.VectorStore = exports.PineconeStore = exports.Chroma = exports.HNSWLib = void 0;
/* #__PURE__ */ console.error("[WARN] Importing from 'langchain/vectorstores' is deprecated. Import from eg. 'langchain/vectorstores/pinecone' instead. See https://js.langchain.com/docs/getting-started/install#updating-from-0052 for upgrade instructions.");
var hnswlib_js_1 = require("./hnswlib.cjs");
Object.defineProperty(exports, "HNSWLib", { enumerable: true, get: function () { return hnswlib_js_1.HNSWLib; } });
var chroma_js_1 = require("./chroma.cjs");
Object.defineProperty(exports, "Chroma", { enumerable: true, get: function () { return chroma_js_1.Chroma; } });
var pinecone_js_1 = require("./pinecone.cjs");
Object.defineProperty(exports, "PineconeStore", { enumerable: true, get: function () { return pinecone_js_1.PineconeStore; } });
var base_js_1 = require("./base.cjs");
Object.defineProperty(exports, "VectorStore", { enumerable: true, get: function () { return base_js_1.VectorStore; } });
Object.defineProperty(exports, "SaveableVectorStore", { enumerable: true, get: function () { return base_js_1.SaveableVectorStore; } });
var supabase_js_1 = require("./supabase.cjs");
Object.defineProperty(exports, "SupabaseVectorStore", { enumerable: true, get: function () { return supabase_js_1.SupabaseVectorStore; } });
var prisma_js_1 = require("./prisma.cjs");
Object.defineProperty(exports, "PrismaVectorStore", { enumerable: true, get: function () { return prisma_js_1.PrismaVectorStore; } });
