// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/models/CreateRequest.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Pinecone API
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: version not set
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.CreateRequestToJSON = exports.CreateRequestFromJSONTyped = exports.CreateRequestFromJSON = exports.instanceOfCreateRequest = void 0;
var runtime_1 = require("../runtime");
var CreateRequestIndexConfig_1 = require("./CreateRequestIndexConfig");
/**
 * Check if a given object implements the CreateRequest interface.
 */
function instanceOfCreateRequest(value) {
    var isInstance = true;
    isInstance = isInstance && "name" in value;
    isInstance = isInstance && "dimension" in value;
    return isInstance;
}
exports.instanceOfCreateRequest = instanceOfCreateRequest;
function CreateRequestFromJSON(json) {
    return CreateRequestFromJSONTyped(json, false);
}
exports.CreateRequestFromJSON = CreateRequestFromJSON;
function CreateRequestFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'name': json['name'],
        'dimension': json['dimension'],
        'indexType': !(0, runtime_1.exists)(json, 'index_type') ? undefined : json['index_type'],
        'metric': !(0, runtime_1.exists)(json, 'metric') ? undefined : json['metric'],
        'pods': !(0, runtime_1.exists)(json, 'pods') ? undefined : json['pods'],
        'replicas': !(0, runtime_1.exists)(json, 'replicas') ? undefined : json['replicas'],
        'shards': !(0, runtime_1.exists)(json, 'shards') ? undefined : json['shards'],
        'podType': !(0, runtime_1.exists)(json, 'pod_type') ? undefined : json['pod_type'],
        'indexConfig': !(0, runtime_1.exists)(json, 'index_config') ? undefined : (0, CreateRequestIndexConfig_1.CreateRequestIndexConfigFromJSON)(json['index_config']),
        'metadataConfig': !(0, runtime_1.exists)(json, 'metadata_config') ? undefined : json['metadata_config'],
        'sourceCollection': !(0, runtime_1.exists)(json, 'source_collection') ? undefined : json['source_collection'],
    };
}
exports.CreateRequestFromJSONTyped = CreateRequestFromJSONTyped;
function CreateRequestToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'name': value.name,
        'dimension': value.dimension,
        'index_type': value.indexType,
        'metric': value.metric,
        'pods': value.pods,
        'replicas': value.replicas,
        'shards': value.shards,
        'pod_type': value.podType,
        'index_config': (0, CreateRequestIndexConfig_1.CreateRequestIndexConfigToJSON)(value.indexConfig),
        'metadata_config': value.metadataConfig,
        'source_collection': value.sourceCollection,
    };
}
exports.CreateRequestToJSON = CreateRequestToJSON;
