// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/models/IndexMetaDatabaseStatus.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Pinecone API
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: version not set
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.IndexMetaDatabaseStatusToJSON = exports.IndexMetaDatabaseStatusFromJSONTyped = exports.IndexMetaDatabaseStatusFromJSON = exports.instanceOfIndexMetaDatabaseStatus = exports.IndexMetaDatabaseStatusStateEnum = void 0;
var runtime_1 = require("../runtime");
/**
 * @export
 */
exports.IndexMetaDatabaseStatusStateEnum = {
    Initializing: 'Initializing',
    ScalingUp: 'ScalingUp',
    ScalingDown: 'ScalingDown',
    Terminating: 'Terminating',
    Ready: 'Ready'
};
/**
 * Check if a given object implements the IndexMetaDatabaseStatus interface.
 */
function instanceOfIndexMetaDatabaseStatus(value) {
    var isInstance = true;
    return isInstance;
}
exports.instanceOfIndexMetaDatabaseStatus = instanceOfIndexMetaDatabaseStatus;
function IndexMetaDatabaseStatusFromJSON(json) {
    return IndexMetaDatabaseStatusFromJSONTyped(json, false);
}
exports.IndexMetaDatabaseStatusFromJSON = IndexMetaDatabaseStatusFromJSON;
function IndexMetaDatabaseStatusFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'ready': !(0, runtime_1.exists)(json, 'ready') ? undefined : json['ready'],
        'state': !(0, runtime_1.exists)(json, 'state') ? undefined : json['state'],
    };
}
exports.IndexMetaDatabaseStatusFromJSONTyped = IndexMetaDatabaseStatusFromJSONTyped;
function IndexMetaDatabaseStatusToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'ready': value.ready,
        'state': value.state,
    };
}
exports.IndexMetaDatabaseStatusToJSON = IndexMetaDatabaseStatusToJSON;
