// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/models/IndexMetaDatabase.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Pinecone API
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: version not set
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.IndexMetaDatabaseToJSON = exports.IndexMetaDatabaseFromJSONTyped = exports.IndexMetaDatabaseFromJSON = exports.instanceOfIndexMetaDatabase = void 0;
var runtime_1 = require("../runtime");
var IndexMetaDatabaseIndexConfig_1 = require("./IndexMetaDatabaseIndexConfig");
/**
 * Check if a given object implements the IndexMetaDatabase interface.
 */
function instanceOfIndexMetaDatabase(value) {
    var isInstance = true;
    return isInstance;
}
exports.instanceOfIndexMetaDatabase = instanceOfIndexMetaDatabase;
function IndexMetaDatabaseFromJSON(json) {
    return IndexMetaDatabaseFromJSONTyped(json, false);
}
exports.IndexMetaDatabaseFromJSON = IndexMetaDatabaseFromJSON;
function IndexMetaDatabaseFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'name': !(0, runtime_1.exists)(json, 'name') ? undefined : json['name'],
        'dimensions': !(0, runtime_1.exists)(json, 'dimensions') ? undefined : json['dimensions'],
        'indexType': !(0, runtime_1.exists)(json, 'index_type') ? undefined : json['index_type'],
        'metric': !(0, runtime_1.exists)(json, 'metric') ? undefined : json['metric'],
        'pods': !(0, runtime_1.exists)(json, 'pods') ? undefined : json['pods'],
        'replicas': !(0, runtime_1.exists)(json, 'replicas') ? undefined : json['replicas'],
        'shards': !(0, runtime_1.exists)(json, 'shards') ? undefined : json['shards'],
        'podType': !(0, runtime_1.exists)(json, 'pod_type') ? undefined : json['pod_type'],
        'indexConfig': !(0, runtime_1.exists)(json, 'index_config') ? undefined : (0, IndexMetaDatabaseIndexConfig_1.IndexMetaDatabaseIndexConfigFromJSON)(json['index_config']),
        'metadataConfig': !(0, runtime_1.exists)(json, 'metadata_config') ? undefined : json['metadata_config'],
    };
}
exports.IndexMetaDatabaseFromJSONTyped = IndexMetaDatabaseFromJSONTyped;
function IndexMetaDatabaseToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'name': value.name,
        'dimensions': value.dimensions,
        'index_type': value.indexType,
        'metric': value.metric,
        'pods': value.pods,
        'replicas': value.replicas,
        'shards': value.shards,
        'pod_type': value.podType,
        'index_config': (0, IndexMetaDatabaseIndexConfig_1.IndexMetaDatabaseIndexConfigToJSON)(value.indexConfig),
        'metadata_config': value.metadataConfig,
    };
}
exports.IndexMetaDatabaseToJSON = IndexMetaDatabaseToJSON;
