// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/models/PatchRequest.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Pinecone API
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: version not set
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.PatchRequestToJSON = exports.PatchRequestFromJSONTyped = exports.PatchRequestFromJSON = exports.instanceOfPatchRequest = void 0;
var runtime_1 = require("../runtime");
/**
 * Check if a given object implements the PatchRequest interface.
 */
function instanceOfPatchRequest(value) {
    var isInstance = true;
    return isInstance;
}
exports.instanceOfPatchRequest = instanceOfPatchRequest;
function PatchRequestFromJSON(json) {
    return PatchRequestFromJSONTyped(json, false);
}
exports.PatchRequestFromJSON = PatchRequestFromJSON;
function PatchRequestFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'replicas': !(0, runtime_1.exists)(json, 'replicas') ? undefined : json['replicas'],
        'podType': !(0, runtime_1.exists)(json, 'pod_type') ? undefined : json['pod_type'],
    };
}
exports.PatchRequestFromJSONTyped = PatchRequestFromJSONTyped;
function PatchRequestToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'replicas': value.replicas,
        'pod_type': value.podType,
    };
}
exports.PatchRequestToJSON = PatchRequestToJSON;
